torch/distributed/distributed_c10d.py

```diff
@@ -295,9 +295,7 @@ def _get_group_size(group):
     if group is GroupMember.WORLD or group is None:
         default_pg = _get_default_group()
         return default_pg.size()
-    if group not in _pg_group_ranks:
-        raise RuntimeError("The given group does not exist")
-    return len(_pg_group_ranks[group])
+    return group.size()
```

Comment:
This change (and the one below) were failing because `all_gather_object` was failing when used with a ProcessGroup instance which was created "by hand", because it didn't have an entry in the global variables `_pg_map` and `_pg_group_ranks`. I don't think there's any fundamental reason for that restriction, hence here I attempted to fix it. Not sure if this is the best way to go about it? CC @rohan-varma

Reply:
I think this is fine for now, although ideally we shouldn't have to modify these private APIs to get things to work properly as this is basically a process group that c10d is not "aware of" in some sense.

On the other hand, it looks like rpc's `_init_process_group` needs to do this because we don't want to initialize a default group. One alternate solution here is to register that pg in the private maps, although that's not great either. We could consider just initializing the default group in that case, though that comes with tradeoffs as well.

Reply:
What I don't understand is why a manually-created process group is not self-sufficient? Why won't it fully work unless some additional metadata is store _somewhere else_? If this metadata is always needed, shouldn't we store it on the ProcessGroup object itself? If it's never needed, should we drop it? If it's only sometimes needed, we should figure out where that's strictly the case, and only make those operations fail, but all other operations should work.

Reply:
@lw I think this is a relic from the original design of c10d python package, where a python-side process group is not considered properly registered unless it is created with `init_process_group` or `new_group` APIs. i.e. there is no guaranteed support of process groups created directly from the pybind APIs. Torchrec team ran into a bunch of issues around this internally earlier as well. 

Why this is the case will probably require digging into the original design of c10d python package, but agreed that it would be nice to support this. As a first step we can probably try to document all APIs that break if the process group is manually created using the pybind APIs.